### PR TITLE
feat(x/ecocredit): add minimal basket proto

### PR DESCRIPTION
**Pull request by aaronc**
## Description

Minimal basket proto to align with super quick Regen Ledger 2.1 release. To be backported against v2.0.x.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/regen/ecocredit/basket/v1alpha1/state.proto
Comment (technicallyty) [inline proto/regen/ecocredit/basket/v1alpha1/state.proto:23]: 1NF?

Comment (aaronc) [reply proto/regen/ecocredit/basket/v1alpha1/state.proto:23]: right

File: proto/regen/ecocredit/basket/v1alpha1/state.proto
Comment (technicallyty) [inline proto/regen/ecocredit/basket/v1alpha1/state.proto:25]:
```suggestion
    // credit_type_name is the name of the credit type this basket is able to hold.
```

File: proto/regen/ecocredit/basket/v1alpha1/state.proto
Comment (technicallyty) [inline proto/regen/ecocredit/basket/v1alpha1/state.proto:22]:
```suggestion
    // disable_auto_retire indicates whether or not the credits will be retired upon withdraw from the basket.
```

File: proto/regen/ecocredit/basket/v1alpha1/state.proto
Comment (technicallyty) [inline proto/regen/ecocredit/basket/v1alpha1/state.proto:31]:
```suggestion
    // multiplier is an integer number which is applied to credit units when
    // converting to basket units. For example if the multiplier is 2000, then
    // 1.1 credits will result in 2200 basket tokens. If there are any fractional
    // amounts left over in this calculation when adding credits to a basket,
    // those fractional amounts will not get added to the basket.
    string multiplier = 6;
```

File: proto/regen/ecocredit/basket/v1alpha1/state.proto
Comment (technicallyty) [inline proto/regen/ecocredit/basket/v1alpha1/state.proto:41]:
```suggestion
    // basket_denom is the denom of the basket
    string basket_denom = 1;
```

File: proto/regen/ecocredit/basket/v1alpha1/state.proto
Comment (technicallyty) [inline proto/regen/ecocredit/basket/v1alpha1/state.proto:42]:
```suggestion
    // class_id is the id of the credit class that is allowed to be deposited in the basket
    string class_id = 2;
```

File: proto/regen/ecocredit/basket/v1alpha1/state.proto
Comment (technicallyty) [inline proto/regen/ecocredit/basket/v1alpha1/state.proto:62]:
```suggestion
    // batch_start_date is the start date of the batch
    google.protobuf.Timestamp batch_start_date = 4;
```

File: proto/regen/ecocredit/basket/v1alpha1/state.proto
Comment (technicallyty) [inline proto/regen/ecocredit/basket/v1alpha1/state.proto:28]:
```suggestion
  // batch_date_window is a duration of time measured into the past which sets
  // a cutoff for batch start dates when adding new credits to the basket.
  // Based on the current block timestamp, credits whose start date is before
  // block timestamp - batch_date_window will not be allowed into the basket.
```

File: proto/regen/ecocredit/basket/v1alpha1/tx.proto
Comment (technicallyty) [inline proto/regen/ecocredit/basket/v1alpha1/tx.proto:53]:
```suggestion
   // allowed_classes are the credit classes allowed to be put in the basket
  repeated string allowed_classes = 7;
```

File: proto/regen/ecocredit/basket/v1alpha1/tx.proto
Comment (technicallyty) [inline proto/regen/ecocredit/basket/v1alpha1/tx.proto:31]:
```suggestion
  // name will be used to create a bank denom for this basket token
```
omit, `:` isnt valid in denoms

File: proto/regen/ecocredit/basket/v1alpha1/tx.proto
Comment (technicallyty) [inline proto/regen/ecocredit/basket/v1alpha1/tx.proto:35]:
```suggestion
  // display_name will be used to create a bank Metadata display name for this
  // basket token
```
ditto

**Review by technicallyty — APPROVED** [review]
lgtm, just comment nits

File: proto/regen/ecocredit/basket/v1alpha1/tx.proto
Comment (aaronc) [reply proto/regen/ecocredit/basket/v1alpha1/tx.proto:31]: I had a PR in the SDK to change the regex. Did it not get merged?

Comment (technicallyty) [reply proto/regen/ecocredit/basket/v1alpha1/tx.proto]: perhaps is not in our fork? just tested on `github.com/regen-network/cosmos-sdk v0.44.2-regen-1` and it fails

File: proto/regen/ecocredit/basket/v1alpha1/tx.proto
Comment (clevinson) [inline proto/regen/ecocredit/basket/v1alpha1/tx.proto:110]:
```suggestion
  string retirement_location = 4;
  
  // retire_on_take is a boolean that dictates whether the ecocredits
  // received in exchange for the basket tokens will be received as
  // retired or tradable credits.
  bool retire_on_take = 5;
```

File: proto/regen/ecocredit/basket/v1alpha1/tx.proto
Comment (clevinson) [inline proto/regen/ecocredit/basket/v1alpha1/tx.proto:117]: `BasketCredit`s only have a field for "tradable_amount", but sometimes these credits may actually be retired (if `retired_on_take` is true).

Seems like the easiest solution is maybe to change the `BasketCredit` message to have the field named `amount`instead of `tradable_amount`.

File: proto/regen/ecocredit/basket/v1alpha1/types.proto
Comment (clevinson) [inline proto/regen/ecocredit/basket/v1alpha1/types.proto:16]:
```suggestion
  // amount is the number of credits associated with the given `batch_denom`
  // Decimal values are acceptable within the precision of the corresponding
  // credit type for this batch.
  string amount = 2;
```

Suggesting changing `tradable_amount` => `amount` as these credits may be retired when sent back to the user if `retire_on_take` is true.

**Review by clevinson — COMMENTED** [review]
Looking good! 

Few small suggestions/questions that I think we should address before merging. Nothing major.

File: proto/regen/ecocredit/basket/v1alpha1/state.proto
Comment (aaronc) [reply proto/regen/ecocredit/basket/v1alpha1/state.proto:16]: it's not really backporting, but yes we'd be using it

**Comment by aaronc** [comment]
Pushed several updates. Please re-review @clevinson @technicallyty

File: proto/regen/ecocredit/basket/v1/state.proto
Comment (robert-zaremba) [inline proto/regen/ecocredit/basket/v1/state.proto:22]:
```suggestion
    // basket_denom is the basket bank denom.
```

File: proto/regen/ecocredit/basket/v1/state.proto
Comment (robert-zaremba) [inline proto/regen/ecocredit/basket/v1/state.proto:33]:
```suggestion
    google.protobuf.Timestamp start_date = 5;
```

Comment (robert-zaremba) [reply proto/regen/ecocredit/basket/v1/state.proto:33]: Let's remove `min`, I think `start_date` is enough.

File: proto/regen/ecocredit/basket/v1/tx.proto
Comment (robert-zaremba) [inline proto/regen/ecocredit/basket/v1/tx.proto:66]: We need to save it in case a `Params.basket_creation_fee` will change, right?

File: proto/regen/ecocredit/basket/v1/tx.proto
Comment (robert-zaremba) [inline proto/regen/ecocredit/basket/v1/tx.proto:84]:
```suggestion
  string sender = 1;
```

**Review by robert-zaremba — APPROVED** [review]
not blocking, left few suggestions. I wonder if we should use `basket.id` instead of `denom` in all the mappings.

File: proto/regen/ecocredit/basket/v1/state.proto
Comment (ryanchristo) [reply proto/regen/ecocredit/basket/v1/state.proto:33]: I would prefer to keep `min` but no strong preference.

File: proto/regen/ecocredit/basket/v1/state.proto
Comment (ryanchristo) [inline proto/regen/ecocredit/basket/v1/state.proto:70]: Why do we need this field for basket balance?

File: proto/regen/ecocredit/basket/v1/tx.proto
Comment (ryanchristo) [reply proto/regen/ecocredit/basket/v1/tx.proto:84]: I would prefer `owner` but no strong preference.

**Review by ryanchristo — APPROVED** [review]
Looks good to me. A couple comments but nothing blocking.

File: proto/regen/ecocredit/basket/v1/tx.proto
Comment (aaronc) [reply proto/regen/ecocredit/basket/v1/tx.proto:66]: What do you mean?

File: proto/regen/ecocredit/basket/v1/state.proto
Comment (aaronc) [reply proto/regen/ecocredit/basket/v1/state.proto:33]: let's keep `min_`. it's not really start date but the minimum or earliest start date. could be `earliest_start_date` but that's even longer

File: proto/regen/ecocredit/basket/v1/tx.proto
Comment (aaronc) [reply proto/regen/ecocredit/basket/v1/tx.proto:84]: hmm.. no strong preference either, but easier to keep owner

File: proto/regen/ecocredit/basket/v1/state.proto
Comment (aaronc) [reply proto/regen/ecocredit/basket/v1/state.proto:70]: to remove oldest credits first. will add a comment

File: proto/regen/ecocredit/basket/v1/tx.proto
Comment (robert-zaremba) [reply proto/regen/ecocredit/basket/v1/tx.proto:66]: I was wondering why do we need to save the fee, rather than just charging the minimum.

Comment (aaronc) [reply proto/regen/ecocredit/basket/v1/tx.proto]: it's for signing UX so the user isn't surprised by the fee. added a comment

**Comment by aaronc** [comment]
auto-merging to unblock the next pieces

File: proto/regen/ecocredit/basket/v1/tx.proto
Comment (robert-zaremba) [reply proto/regen/ecocredit/basket/v1/tx.proto:87]: In Ethereum that would be `sender`

**Pull request merged**
